### PR TITLE
Add `envWithCleanup`

### PR DESCRIPTION
**Pull request by tvh**
This is a fix for #79

### CODE REVIEW

File: Criterion/Internal.hs [hunk 1]
```diff
@@ -182,13 +182,15 @@ for :: MonadIO m => (String -> Bool) -> Benchmark
...
-    go !idx (pfx, Environment mkenv mkbench)
+    go !idx (pfx, Environment mkenv cleanupenv mkbench)
       | shouldRun pfx mkbench = do
         e <- liftIO $ do
           ee <- mkenv
           evaluate (rnf ee)
           return ee
-        go idx (pfx, mkbench e)
+        idx' <- go idx (pfx, mkbench e)
+        liftIO $ cleanupenv e
```
Comment (mgsloan) [inline Criterion/Internal.hs]: I don't think that is quite correct.  You want something like `finally (go idx (pfx, mkbench e)) (liftIO (cleanupenv e))`

**Comment by mgsloan** [comment]
Thanks for working on this!  This isn't correct for exceptions.  The exception case is the tricky one, and the entire reason an explicit cleanup action is required.

A test should also be added to make sure it works with exceptions.

**Comment by tvh** [comment]
Sorry for taking this long to get back at this.

I know this is nor correct for exceptions, but I assumed it is okay, as

**Comment by RyanGlScott** [comment]
This has been superseded by #136.

**Pull request closed without merging**